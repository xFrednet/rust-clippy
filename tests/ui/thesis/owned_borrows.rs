//@rustc-env: CLIPPY_PETS_PRINT=1

struct Animal {
    science_name: String,
    simple_name: String,
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_1(a: String) -> bool {
    a.is_empty()
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_2(a: String) {
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_3(a: String) {
    take_2_loan(&a, &a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_1(mut a: String) {
    a.clear();
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_2(mut a: String) {
    take_1_mut_loan(&mut a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_3(mut a: Animal) {
    take_2_mut_loan(&mut a.science_name, &mut a.simple_name);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed(mut a: String) {
    take_1_mut_loan(&mut a);
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed_2(mut a: Animal) {
    take_2_mixed_loan(&a.science_name, &mut a.simple_name);
}

/// https://github.com/nikomatsakis/nll-rfc/issues/37
#[warn(clippy::borrow_pats)]
fn two_phase_borrow_1(mut vec: Vec<usize>) {
    vec.push(vec.len());
}

#[warn(clippy::borrow_pats)]
fn two_phase_borrow_2(mut num: usize, mut vec: Vec<usize>) {
    vec.push({
        num = vec.len();
        num
    })
}

fn take_1_loan(_: &String) {}
fn take_2_loan(_: &String, _: &String) {}
fn take_1_mut_loan(_: &String) {}
fn take_2_mut_loan(_: &mut String, _: &mut String) {}
fn take_2_mixed_loan(_: &String, _: &mut String) {}

fn main() {}
