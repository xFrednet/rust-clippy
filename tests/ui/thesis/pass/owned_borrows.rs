//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_STATS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

#[derive(Default)]
struct Animal {
    science_name: String,
    simple_name: String,
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_1(a: String) -> bool {
    a.is_empty()
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_2(a: String) {
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_3(a: String) {
    take_2_loan(&a, &a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_1(mut a: String) {
    a.clear();
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_2(mut a: String) {
    take_1_mut_loan(&mut a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mut_3(mut a: Animal) {
    take_2_mut_loan(&mut a.science_name, &mut a.simple_name);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed(mut a: String) {
    take_1_mut_loan(&mut a);
    take_1_loan(&a);
}

#[warn(clippy::borrow_pats)]
fn temp_borrow_mixed_2(mut a: Animal) {
    take_2_mixed_loan(&a.science_name, &mut a.simple_name);
}

/// https://github.com/nikomatsakis/nll-rfc/issues/37
#[warn(clippy::borrow_pats)]
fn two_phase_borrow_1(mut vec: Vec<usize>) {
    vec.push(vec.len());
}

#[warn(clippy::borrow_pats)]
fn two_phase_borrow_2(mut num: usize, mut vec: Vec<usize>) {
    vec.push({
        num = vec.len();
        num
    })
}

struct NestedVecs {
    a: Vec<usize>,
    b: Vec<usize>,
}

#[warn(clippy::borrow_pats)]
fn nested_two_phase_borrow(mut vecs: NestedVecs) {
    vecs.a.push({
        vecs.b.push(vecs.a.len());
        vecs.b.len()
    });
}

#[warn(clippy::borrow_pats)]
fn test_double_loan() {
    let data = "Side effects".to_string();
    take_double_loan(&&data);
}

#[warn(clippy::borrow_pats)]
fn test_double_mut_loan() {
    let mut data = "Can Side effects".to_string();
    take_double_mut_loan(&&mut data);
}

#[warn(clippy::borrow_pats)]
fn test_mut_double_loan() {
    let data = "Can't really have Side effects".to_string();
    take_mut_double_loan(&mut &data);
}

#[forbid(clippy::borrow_pats)]
fn loan_to_access_part() {
    let data = Animal::default();
    take_1_loan(&(&&data).simple_name);
}

fn take_1_loan(_: &String) {}
fn take_2_loan(_: &String, _: &String) {}
fn take_1_mut_loan(_: &String) {}
fn take_2_mut_loan(_: &mut String, _: &mut String) {}
fn take_2_mixed_loan(_: &String, _: &mut String) {}
fn take_double_loan(_: &&String) {}
fn take_double_mut_loan(_: &&mut String) {}
fn take_mut_double_loan(_: &mut &String) {}

fn main() {}
