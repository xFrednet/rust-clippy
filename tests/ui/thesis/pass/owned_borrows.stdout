# "temp_borrow_1"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Primitive), NotConst , Safe  , Sync , Free       {HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_2"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_3"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, AliasedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mut_1"
- a              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mut_2"
- a              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mut_3"
- a              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {PartBorrow, PartArgBorrowMut, MultipleMutBorrowsInArgs}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 2,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mixed"
- a              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mixed_2"
- a              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {PartBorrow, PartArgBorrow, PartArgBorrowMut, MixedBorrowsInArgs}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "two_phase_borrow_1"
- vec            : (Mutable  , Owned    , Argument, NonCopy, SelfDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut, TwoPhasedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut, HasTwoPhaseBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 1,
    },
}

# "two_phase_borrow_2"
- num            : (Mutable  , Owned    , Argument, Copy   , NonDrop , Primitive) {Overwrite}
- vec            : (Mutable  , Owned    , Argument, NonCopy, SelfDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut, TwoPhasedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut, HasTwoPhaseBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 1,
    },
}

# "nested_two_phase_borrow"
- vecs           : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {PartBorrow, PartArgBorrow, PartArgBorrowMut, TwoPhasedBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut, HasTwoPhaseBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 2,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 1,
    },
}

# "test_double_loan"
- data           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "test_double_mut_loan"
- data           : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "test_mut_double_loan"
- data           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "loan_to_access_part"
=====
Body for: DefId(0:20 ~ owned_borrows[a1d0]::loan_to_access_part)
bb0:
    StorageLive(_1)
    _1 = <Animal as std::default::Default>::default() -> [return: bb1, unwind continue]

bb1:
    StorageLive(_2)
    StorageLive(_3)
    StorageLive(_4)
    StorageLive(_5)
    StorageLive(_6)
    _6 = &_1
    _5 = &_6
    _7 = deref_copy (*_5)
    _4 = &((*_7).1: std::string::String)
    _3 = &(*_4)
    _2 = take_1_loan(move _3) -> [return: bb2, unwind: bb4]

bb2:
    StorageDead(_3)
    StorageDead(_6)
    StorageDead(_5)
    StorageDead(_4)
    StorageDead(_2)
    _0 = const ()
    drop(_1) -> [return: bb3, unwind: bb5]

bb3:
    StorageDead(_1)
    return

bb4:
    drop(_1) -> [return: bb5, unwind terminate(cleanup)]

bb5:
    resume

=====
- data           : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 1,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

